op-interop-mon: replace `BlockBuffer` with generic wrapper for `container/ring.Ring` (std)

geoknee opened the pull request:
* Replaces `BlockBuffer` implementation, but maintains API (mostly)
* Uses standard library ring structure
* Wraps with generics to adorn ring with type information
* Adds unit tests, which can use lightweight test data thanks to generic type
* installs into op-service.

geoknee committed:
WIP

geoknee committed:
improve

geoknee commented on op-interop-mon/monitor/block_buffer.go:
```diff
@@ -1,56 +1,74 @@
+	return b
+}
+
+func (r *Buffer[T]) Print() {
+	fmt.Println("--------------------------------")
+	r.Do(func(v any) {
+		fmt.Printf("%v\n", format(v))
+	})
+	p := r.Peek()
+	n := r.Next()
+	fmt.Printf("peek: %v\n", p)
+	fmt.Printf("next: %v\n", n.Value)
+	fmt.Println("--------------------------------")
+}
+
+func format(v any) string {
+	switch v := v.(type) {
+	case *int:
+		if v == nil {
+			return "nil"
+		}
+		return fmt.Sprintf("%v", *v)
+	default:
+		return fmt.Sprintf("%v", v)
```
This will be removed shortly and is just helping me debug the impl.

geoknee committed:
cleanup

geoknee committed:
rename

geoknee committed:
more rename and tidy

geoknee commented on op-interop-mon/monitor/buffer.go:
```diff
@@ -0,0 +1,52 @@
+// A RingBuffer is a generic [container/ring.Ring]
+// with convenient methods for adding and removing values.
+// The generic type T should be a pointer type or interface
+// because the default value if T will be returned if the value
```
```suggestion
// because the default value of T will be returned by Pop and Peek if the value
```

geoknee committed:
Update op-interop-mon/monitor/buffer.go

geoknee commented on op-interop-mon/monitor/buffer.go:
```diff
@@ -0,0 +1,52 @@
+// because the default value of T will be returned by Pop and Peek if the value
+// is unset.
+type RingBuffer[T any] struct {
+	*ring.Ring
```
It might be better to not embed this, and then we can prevent the caller from reaching past the API in this file and accessing the `any` Value of the `ring.Ring` type.

geoknee replied:
Went ahead and did this [1bdb2cf](https://github.com/ethereum-optimism/optimism/pull/16382/commits/1bdb2cf36f3a2f76cf032912b033012328b4b87d)

axelKingsley reviewed (commented):
At first glance this looks lovely -- maybe a useful thing to move to op-service?

geoknee committed:
move ring buffer type to op-service

geoknee committed:
op-service/buffer: do not embed implementation struct

this is safe because the caller cannot reach into the non-type-safe values stored within

geoknee commented:
> At first glance this looks lovely -- maybe a useful thing to move to op-service?

Went ahead and pushed it into op-service.

geoknee commented on op-interop-mon/monitor/finder.go:
```diff
@@ -178,10 +160,10 @@ func (t *RPCFinder) processBlock(blockInfo eth.BlockInfo, receipts types.Receipt
-		if err != nil {
-			t.log.Error("error popping block", "error", err)
-			return err
+		previous := t.seenBlocks.Pop()
+		if previous == nil {
+			t.log.Error("no blocks to walk back to")
+			return ErrBlockNotFound
```
I think we will want to cover this case in the `finder_test.go`. It would happen if there was a reorg deeper than the seen blocks buffer length. 

While we are at that, it would be good to make the buffer length more configurable (perhaps the finder constructor accepts it, it could even be a CLI flag).

axelKingsley replied:
Yeah agreed, configurability is still sorely needed across the mon. and good callout re test. I don't think this needs to be handled in this PR tho

axelKingsley commented on op-service/buffer/buffer.go:
```diff
@@ -0,0 +1,56 @@
+	c *ring.Ring
+}
+
+// NewBlockBuffer creates a new RingBuffer
```
```suggestion
// NewRing creates a new RingBuffer
```

axelKingsley commented on op-service/buffer/buffer.go:
```diff
@@ -0,0 +1,56 @@
+	return &b
+}
+
+// Add adds a value to the RingBuffer, removing the oldest value
```
```suggestion
// Add adds a value to the RingBuffer, overwriting the oldest value if full
```

axelKingsley reviewed (approved):
Looks good, a couple comment updates to be had.

We discussed offline, but if possible let's merge the finality expiry first and then I can merge this on top

axelKingsley committed:
Merge remote-tracking branch 'origin' into gk/ring-buffer

axelKingsley committed:
comment fixup
